### PR TITLE
Support quoting parameter names

### PR DESCRIPTION
**Pull request by mattmoor**
# Changes

/kind feature

This change plumbs through support for quoting user-provided names during variable resolution as proposed [here](https://github.com/tektoncd/pipeline/issues/3590#issuecomment-744576896).

In addition to `$(params.foo)` you can now use `$(params."foo")`, which is useful for disambiguation when the name contains `.` characters (e.g. `$(params."dev.tekton.foo")`).

I have updated unit tests, and added a minor update to docs, but if there are more tests or docs that should be updated, please send me pointers.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [ ] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
Support quoting user-provided names during variable substitution, e.g. $(params."foo") or $(results."bar".path)
```

### CODE REVIEW

**Comment by mattmoor** [comment]
cc @vdemeester @afrittoli @ImJasonH

**Comment by mattmoor** [comment]
Something strange is going on, both of those passed.  I think in the past this correlated with Prow being overloaded?

**Comment by sbwsg** [comment]
This feature addition feels a tad under-specified to me.  Is this intended as a proposal or do you want us to proceed on merging this as-is?

Here's what I'm gathering from reading the code: This change affects params, workspaces, results and resource variables.  From an initial scan it doesn't look like we've added this support to `context` variables, for example.  Only certain portions of variables are quotable (e.g. `"resources.%s.%q.%s"` is supported but not `"resources.%q.%s.%s"`; `"workspaces.%q.%s"` is supported but not `"workspaces.%q.%q"`).  The doc atm doesn't capture these ambiguities or explain the reasoning.  Some `apply` tests have been updated to remove the `inputs.` prefix, which might mean we're dropping a bit of coverage of backwards-compatible behaviour - unclear to me atm.

I'd like to throw another design in to the ring: support bracket notation instead, e.g. `params["my.param.with.periods"]` which could in turn be taken a step further to array param subscripts ala `params.myparam[1]` or `params["my.array.param.with.periods"]["1"]` or something similar.  This isn't so much to dissuade you from running with your design, but I reckon there's enough conceptual room here to at least describe some alternatives in a PR, say why we discount them, and get some conversation going around it.

TL;DR I'm wary of lgtm'ing this without a bit of discussion first.  It might even be TEP-worthy?  Curious what other folks think @tektoncd/core-maintainers

**Comment by mattmoor** [comment]
> "resources.%s.%q.%s" is supported but not "resources.%q.%s.%s"

The intention was to support quote all user-sourced names, there are a few of these (some I'm adding) where the `%s` is used for built-in strings.  In the example above, my inference was that the first `%s` was generally used for "inputs" or "outputs", which is perhaps incorrect.

I am happy to discuss this further, but nobody was discussing anything in the linked issue.  I'm glad this triggered some 😉 

---

On TEPs, perhaps it is worth framing this as an incidental aspect of the broader idea I think we should pursue, which I have roughly sketched here: https://github.com/tektoncd/pipeline/issues/3590#issuecomment-738051808

Ultimately, without that, I have effectively no energy to pursue this further.
